packages/@aws-cdk/aws-ec2/lib/instance.ts

```diff
@@ -361,6 +412,59 @@ export class Instance extends Resource implements IInstance {
     this.role.addToPolicy(statement);
   }
 
+  /**
+   * Use a CloudFormation Init configuration at instance startup
+   *
+   * This does the following:
+   *
+   * - Attaches the CloudFormation Init metadata to the Instance resource.
+   * - Add commands to the instance UserData to run `cfn-init` and `cfn-signal`.
+   * - Update the instance's CreationPolicy to wait for the `cfn-signal` commands.
+   */
+  public applyCloudFormationInit(init: CloudFormationInit, options: ApplyCloudFormationInitOptions = {}) {
+    const platform = this.osType === OperatingSystemType.WINDOWS ? InitPlatform.WINDOWS : InitPlatform.LINUX;
+    init.attach(this.instance, {
+      platform,
+      instanceRole: this.role,
+      userData: this.userData,
+      configSets: options.configSets,
+      embedFingerprint: options.embedFingerprint,
+      printLog: options.printLog,
+      ignoreFailures: options.ignoreFailures,
+    });
+    this.waitForResourceSignal(options.timeout ?? Duration.minutes(5));
+  }
+
+  /**
+   * Wait for a single additional resource signal
+   *
+   * Add 1 to the current ResourceSignal Count and add the given timeout to the current timeout.
+   *
+   * Use this to pause the CloudFormation deployment to wait for the instances
+   * in the AutoScalingGroup to report successful startup during
+   * creation and updates. The UserData script needs to invoke `cfn-signal`
+   * with a success or failure code after it is done setting up the instance.
+   */
+  public waitForResourceSignal(timeout: Duration) {
```

Comment:
Why is this `public`? Does it make sense to call this outside the context of `applyCloudFormationInit`?

Reply:
I think the use case there would be a customer adding/applying their own user-data that called cfn-signal independent of cfn-init.

Reply:
Did we design for this scenario? Not sure its worth exposing this as a sort of addendum feature. Do we want to commit to this API? for example, maybe its better to accept a `count` argument so that users don'r have to split their timeout calculation and invoke this multiple times.

Reply:
TBH, @rix0rrr was the main architect of this side. I could certainly see the argument that this (and applyCloudFormationInit) start out their lives as private, and we expose them publicly after we have some more feedback.